pythonforandroid/graph.py

```diff
@@ -99,17 +132,129 @@ def find_order(graph):
                 bset.discard(result)
 
 
-def get_recipe_order_and_bootstrap(ctx, names, bs=None):
-    recipes_to_load = set(names)
+def obvious_conflict_checker(ctx, name_tuples, blacklist=[]):
+    """ This is a pre-flight check function that will completely ignore
+        recipe order or choosing an actual value in any of the multiple
+        choice tuples/dependencies, and just do a very basic obvious
+        conflict check.
+    """
+    deps_were_added_by = dict()
+    deps = set()
+
+    # Add dependencies for all recipes:
+    to_be_added = [(name_tuple, None) for name_tuple in name_tuples]
+    while len(to_be_added) > 0:
+        current_to_be_added = list(to_be_added)
+        to_be_added = []
+        for (added_tuple, adding_recipe) in current_to_be_added:
+            assert(type(added_tuple) == tuple)
+            if len(added_tuple) > 1:
+                # No obvious commitment in what to add, ignore it.
+                # (Remember this function only catches obvious issues)
+                continue
+
+            name = added_tuple[0]
+            recipe_conflicts = set()
+            recipe_dependencies = []
+            try:
+                # Get recipe to add and who's ultimately adding it:
+                recipe = Recipe.get_recipe(name, ctx)
+                recipe_conflicts = {c.lower() for c in recipe.conflicts}
+                recipe_dependencies = get_dependency_tuple_list_for_recipe(
+                    recipe, blacklist=blacklist
+                )
+            except ValueError:
+                pass
+            adder_first_recipe_name = adding_recipe or name
+
+            # Collect the conflicts:
+            triggered_conflicts = []
+            for dep_tuple_list in deps:
+                # See if the new deps conflict with things added before:
+                if set(dep_tuple_list).intersection(
+                       recipe_conflicts) == dep_tuple_list:
```

Comment:
Now there's a new bug, can you spot it?
Which means coverage is not quite there yet :grimacing:

Reply:
oh fascinating! I don't think this is possible to be hit because `get_dependency_tuple_list_for_recipe` applies `fix_deplist` which ensures all instances of `dep_tuple_list` would already be a set. (so I don't think this is actually possible to trigger via a test?) but yeah, it is kind of nonsensical how it is written now, I'll fix it

Reply:
oh nevermind they're `tuple`s not `set`s so it is a true bug :smile:

Reply:
Okay, I think I might be able to write a test case that hits this. let me give it a try (disclaimer: I'm a bit sleep deprived right now so I'm not promising anything :joy: )

Reply:
I was more talking about the type mismatch `tuple` the other one is `set`
I'm happy unit tests and coverage test can prove some point even for a "simple" function like that one :wink:

Reply:
@AndreMiras wrote a test `test_multichoice_obvious_conflict_checker` that actually hits the bug (gives error before the fix, happy after!) thanks :+1:

Reply:
resolving since hopefully fixed :slightly_smiling_face: